flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMaster.java

```diff
@@ -937,6 +944,34 @@ public CompletableFuture<Acknowledge> notifyNewBlockedNodes(Collection<BlockedNo
         return CompletableFuture.completedFuture(Acknowledge.get());
     }
 
+    @Override
+    public CompletableFuture<JobResourceRequirements> requestJobResourceRequirements() {
+        return CompletableFuture.completedFuture(schedulerNG.requestJobResourceRequirements());
+    }
+
+    @Override
+    public CompletableFuture<Acknowledge> updateJobResourceRequirements(
+            JobResourceRequirements jobResourceRequirements) {
+        final Map<JobVertexID, Integer> maxParallelismPerJobVertex = new HashMap<>();
+        for (JobVertex vertex : jobGraph.getVertices()) {
+            maxParallelismPerJobVertex.put(
+                    vertex.getID(), SchedulerBase.getDefaultMaxParallelism(vertex));
+        }
+        final List<String> validationErrors =
+                JobResourceRequirements.validate(
+                        jobResourceRequirements, maxParallelismPerJobVertex);
+        if (validationErrors.isEmpty()) {
+            schedulerNG.updateJobResourceRequirements(jobResourceRequirements);
```

Comment:
We will handle this being unsupported by the DefaultScheduler directly in the REST layer.
We won't even call this method if the adaptive scheduler isn't used.

Reply:
What are the benefits of handling this error in DefaultScheduler over here?
Something like 
```
        try {
            return CompletableFuture.completedFuture(schedulerNG.updateJobResourceRequirements(...));
        } catch (UnsupportedOperationException e) {
            return FutureUtils.completedExceptionally(
                new RestHandlerException(e.getMessage(), HttpResponseStatus.NOT_IMPLEMENTED)); // not sure about the code
        }
```
Handling `BAD_REQUEST` and `NOT_IMPLEMENTED` (?) in one place seems more consistent to me.

Reply:
The UI will get some buttons for rescaling, and we'd like to hide it if it's not supported by the current cluster configuration. This means we already check the support in the REST layer.
When we do that it's little additional effort to also change the handlers to immediately fail with an error.
We'd shortcut the process that we know would fail anyway, and centralize the handling of whether this is supported in one place.

If we weren't hiding the buttons I'd agree with you (and in fact I implemented that until I remembered about the button hiding :see_no_evil: )

Reply:
I see, thanks for clarifying.